tests/test_directive_patch.py

```diff
@@ -21,7 +21,8 @@ def test_code_directive(app):
             '   print("hello world")\n')
 
     doctree = restructuredtext.parse(app, text)
-    assert_node(doctree, [nodes.document, nodes.literal_block, 'print("hello world")'])
+    assert_node(doctree, ([nodes.literal_block, 'print("hello world")'],
+                          nodes.substitution_definition))
```

Comment:
Finally, I found the reason why the substituion_definition node appears suddenly here. `tests/roots/test-root/conf.py` contains a substitution definition in its `rst_epilog`. We need to remove it to be these testcases healthy. I'll work on it.

Reply:
I just posted https://github.com/sphinx-doc/sphinx/pull/8295 to fix it. So please update this PR. Then I'll review this again.